veniq/dataset_collection/validation.py

```diff
@@ -0,0 +1,284 @@
+
+        for class_decl in ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION):
+            if class_decl.name != class_name:
+                continue
+            elif class_decl.name == class_name:
+                objects_to_consider = list(class_decl.methods) + list(class_decl.constructors) or []
+                for ast_node in objects_to_consider:
+                    result = RowResult(
+                        output_filename=full_path,
+                        input_filename=row[1]['input_filename'],
+                        class_name='',
+                        method_name='',
+                        start_line_SEMI=-1,
+                        end_line_SEMI=-1,
+                        start_line_dataset=start_line_of_inserted_block,
+                        end_line_dataset=end_line_of_inserted_block,
+                        percent_matched=-1.0,
+                        error_string='',
+                        ncss=0,
+                        matched=False,
+                        failed_cases_in_SEMI_algorithm=False,
+                        no_opportunity_chosen=False,
+                        failed_cases_in_validation_examples=False,
+                    )
+                    if ast_node.name != function_to_analyze:
+                        continue
+                    try:
+                        ast_subtree = ast.get_subtree(ast_node)
+                        opport = find_extraction_opportunities(ast_subtree)
+                        if opport:
+                            find_matched_lines(
+                                ast_node,
+                                ast_subtree,
+                                class_decl,
+                                start_line_of_inserted_block,
+                                end_line_of_inserted_block,
+                                full_path,
+                                opport,
+                                result)
+                        else:
+                            result.no_opportunity_chosen = True
+
+                    except Exception as e:
+                        traceback.print_exc()
+                        result.error_string = str(e)
+                        result.failed_cases_in_SEMI_algorithm = True
+                    finally:
+                        results.append(result)
+
+                    break
+                break
+
+    except Exception as e:
+        traceback.print_exc()
+        result.error_string = str(e)
+        result.failed_cases_in_validation_examples = True
+        results.append(result)
+
+    return results
+
+
+def find_matched_lines(
+        ast_node: ASTNode,
+        ast_subtree: AST,
+        class_decl: ASTNode,
+        start_line_of_inserted_block: int,
+        end_line_of_inserted_block: int,
+        full_path: str,
+        opportunities_list: List[ExtractionOpportunityGroup],
+        result: RowResult) -> None:
+    best_group = opportunities_list[0]
```

Comment:
maybe add parameter (for later) top_k with default 1,
so that ``best_group = opportunities_list[top_k-1]''
?

Reply:
can we do it in a separate request? Because we need to fix already worked code. Moreover we need to add new functionality, since not all opportunities can be compared

Reply:
OK